src/libxtp/sternheimerw.cc

```diff
@@ -0,0 +1,404 @@
+      numint.getGridpoints();
+
+  std::cout << "gridsize= " << numint.getGridSize() << std::endl;
+  // Setting up constants
+  const int& num_occ_lvls = _orbitals.getNumberOfAlphaElectrons();
+  const int& basis_size = _orbitals.getBasisSetSize();
+
+  // Setting up constant matrices and vectors needed for the Sternheimer
+  // equation
+  const Eigen::MatrixXd H = Hamiltonian();
+  const Eigen::MatrixXd S = OverlapMatrix();
+  const Eigen::MatrixXd p = DensityMatrix();
+  Eigen::MatrixXcd V = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+
+  const Eigen::MatrixXd& mo_coefficients = _orbitals.MOCoefficients();
+  const Eigen::VectorXd& mo_energies = _orbitals.MOEnergies();
+
+  // Setting up container for solutions in each gridpoint
+
+  Eigen::MatrixXcd delta_c_p = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+  Eigen::MatrixXcd delta_c_m = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+
+  // Setting up container for LHS and RHS of the system
+  std::vector<Eigen::MatrixXcd> LHS_P;
+  std::vector<Eigen::MatrixXcd> LHS_M;
+  Eigen::VectorXcd RHS = Eigen::VectorXcd::Zero(basis_size);
+
+  // H_new is used if omega=0 to shift the eigenvalues away from 0
+  Eigen::MatrixXcd H_new = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+  double alpha = 2 * (mo_energies(num_occ_lvls) - mo_energies(2));
+
+  double pi = tools::conv::Pi;
+  // Initilazing solution vector and container for solution
+
+  std::vector<Eigen::MatrixXcd> result;
+  Eigen::MatrixXcd delta_n = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+
+  // Iterating over all occupied states and saving the LHS
+  for (int v = 0; v < num_occ_lvls; v++) {
+
+    // Setting up LHS since it is not dependent on r
+    if (w == 0.0) {
+      H_new = H + alpha * S * p.transpose();
+      LHS_P.push_back(SternheimerLHS(H_new, S, mo_energies(v), w, true));
+      LHS_M.push_back(SternheimerLHS(H_new, S, mo_energies(v), w, false));
+    } else {
+      LHS_P.push_back(SternheimerLHS(H, S, mo_energies(v), w, true));
+      LHS_M.push_back(SternheimerLHS(H, S, mo_energies(v), w, false));
+    }
+  }
+  // Iterate over all spacial gridpoints
+  for (int r = 0; r < gridpoints.size(); r++) {
+
+    // std::cout<<(*gridpoints[r])<<std::endl;
+    V = CoulombMatrix(*gridpoints.at(r).second);
+
+    for (int v = 0; v < num_occ_lvls; v++) {
+
+      RHS = SternheimerRHS(S, p, V, mo_coefficients.col(v));
+
+      delta_c_p.col(v) = SternheimerSolve(LHS_P.at(v), RHS);
+      delta_c_m.col(v) = SternheimerSolve(LHS_M.at(v), RHS);
+    }
+
+    for (int v = 0; v < basis_size; v++) {
+      for (int i = 0; i < basis_size; i++) {
+        for (int j = 0; j < basis_size; j++) {
+          delta_n(i, j) = delta_n(i, j) +
+                          2 * mo_coefficients(i, v) * delta_c_p(j, v) +
+                          2 * mo_coefficients(i, v) * delta_c_m(j, v);
+        }
+      }
+    }
+
+    result.push_back(delta_n);
+    delta_n = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+
```

Comment:
if you have to reset a matrix that often indicates that you should use it in a smaller scope, i.e. declare it inside the for loop